core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -617,7 +639,44 @@ protected void doRun() throws Exception {
                 finishAsFailed(e);
                 return;
             }
-            finishAndMoveToReplication(replicationPhase);
+
```

Comment:
I have added your changes but kept the exception type of AbstractRunnable.doRun() as before (Exception, not Throwable).